### Importing Libraries

In [13]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

### Importing Data

In [14]:
# Folder Path

path = r'C:\Users\Windows 10\Documents\04-2023 Instacart Basket Analysis'
print(path)

C:\Users\Windows 10\Documents\04-2023 Instacart Basket Analysis


In [15]:
# Importing 'customers.csv'

df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [16]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [17]:
df_customers.shape

(206209, 10)

### Step 1

#### Download the customer data set and add it to your “Original Data” folder : Done

### Step 2

#### Create a new notebook in your “Scripts” folder for part 1 of this task : Done

### Step 3

#### Import your analysis libraries, as well as your new customer data set as a dataframe : Done

### Step 4

#### Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

#### 4a. Renaming the column name

In [18]:
df_customers.rename(columns = {'First Name': 'first_name', 'Surnam': 'last_name', 'Gender': 'gender', 'STATE': 'state',
                              'Age': 'age', 'n_dependants': 'dependants_number', 'fam_status': 'family_status'}, inplace = True)

In [19]:
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants_number family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

#### 4b. Dropping unnecessary column

In [20]:
df_customers = df_customers.drop(columns = ['first_name', 'last_name'])

In [21]:
df_customers.head()

user_id  gender       state  age date_joined  dependants_number  \
0    26711  Female    Missouri   48    1/1/2017                  3   
1    33890  Female  New Mexico   36    1/1/2017                  0   
2    65803    Male       Idaho   35    1/1/2017                  2   
3   125935  Female        Iowa   40    1/1/2017                  0   
4   130797  Female    Maryland   26    1/1/2017                  1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

In [22]:
df_customers.shape

(206209, 8)

##### Now the total of columns is 8

### Step 5

#### Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

#### 5a. Missing values

In [23]:
df_customers.isnull().sum()

user_id              0
gender               0
state                0
age                  0
date_joined          0
dependants_number    0
family_status        0
income               0
dtype: int64

##### No missing values

#### 5b. Finding duplicates

In [24]:
df_customers[df_customers.duplicated()]

Empty DataFrame
Columns: [user_id, gender, state, age, date_joined, dependants_number, family_status, income]
Index: []

##### No duplicates

#### 5c. Mixed-type data

In [25]:
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

##### No mixed_type data

#### 5d. Convert data-type

In [26]:
df_customers.dtypes

user_id               int64
gender               object
state                object
age                   int64
date_joined          object
dependants_number     int64
family_status        object
income                int64
dtype: object

##### I will change the user_id data type to string because its value is not for calculation

In [27]:
# Changing the data type

df_customers['user_id'] = df_customers['user_id'].astype('str')

In [28]:
df_customers['user_id'].dtype

dtype('O')

In [29]:
df_customers.describe()

age  dependants_number         income
count  206209.000000      206209.000000  206209.000000
mean       49.501646           1.499823   94632.852548
std        18.480962           1.118433   42473.786988
min        18.000000           0.000000   25903.000000
25%        33.000000           0.000000   59874.000000
50%        49.000000           1.000000   93547.000000
75%        66.000000           3.000000  124244.000000
max        81.000000           3.000000  593901.000000

### Step 6

#### Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [30]:
# Importing 'orders_products_merged_updated_2.pkl'

df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))

In [31]:
df_ords_prods_merged.dtypes

order_id                     int64
user_id                      int64
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_price              float64
spending_flag               object
median_prior_orders        float64
order_frequency_flag        object
dtype: object

In [32]:
# Changing the data type

df_ords_prods_merged['user_id'] = df_ords_prods_merged['user_id'].astype('str')

In [33]:
# Checking user_id data type from df_ords_prods_merged dataframe

df_ords_prods_merged['user_id'].dtype

dtype('O')

In [34]:
# Checking user_id data type from df_customers dataframe

df_customers['user_id'].dtype

dtype('O')

In [35]:
# Merging dataframe using user_id as a key

df_ords_prods_cust = df_ords_prods_merged.merge(df_customers, on = 'user_id')

In [36]:
df_ords_prods_cust.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  spending_flag  median_prior_orders  \
0         Soda  ...    Low spender                 20.5   
1         Soda  ...    Low spender                 20.5   
2         Soda  ...    Low spender                 20.5   
3         Soda  ...    Low spender                 20.5   
4         Soda  ...    Low spender                 20.5   

    order_frequency_flag  gender    state age date_joined dependants_number  \
0  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
1  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
2  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
3  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
4  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   

   family_status income  
0        married  40423  
1        married  40423  
2        married  40423  
3        married  40423  
4        married  40423  

[5 rows x 31 columns]

In [37]:
# Checking the shape before and afer merging

df_ords_prods_merged.shape

(32404859, 24)

In [38]:
# Checking the shape before and afer merging

df_customers.shape

(206209, 8)

In [39]:
# Checking the shape before and afer merging

df_ords_prods_cust.shape

(32404859, 31)

##### Now the row total is 32404859 which corresponds to the row total from df_ords_prods_merged and the column total is now 31 (from 24+8). They are not 32 because we use user_id as key to combine both dataframes which they have the same value and no need to duplicate

### Step 7

#### Ensure your notebook contains logical titles, section headings, and descriptive code comments : Done

### Step 8

#### Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [41]:
df_ords_prods_cust.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers.pkl'))